### DISCUSSION TITLE
Cameras supporting more than 30 fps

### DISCUSSION THREAD

**Discussion by Ernovation** [pedroSG94/RootEncoder#1388] [General]
Hi all,

I am using this fantastic library for a video streaming app. I stream video to Facebook or Youtube, after I put an overlay on the video. I use this on a Galaxy S23 Ultra (Samsung SM-S918B).

When I set the frame rate to more than 30 fps, the frame rate is still 30 fps. It appears that my phone does not support more than 30 fps.

I got the characteristics of the different cameras returned by CameraManager.getCameraIdList(). Then I got the StreamConfigurationMap for each camera. These maps for each camera all have a duration of 33333333 ns for each size and format, corresponding to 30 fps. I find this disappointing, since a Galaxy S23 Ultra supports much higher frame rates in the built-in camera app.

Is there way to get more than 30 fps with this phone? Are there other phones that support higher frame rates, also for 3840x2160?

Regards,
*Erik.

**Comment by Ernovation** [comment]
I just checked my Galaxy S10 (Samsung SM-G973F), and it has at least some formats with minFrameDutation of 16665880 ns (including for size 1920x1080), corresponding to 60 fps. Also my Galaxy S21 has formats and sizes with minFrameDuration of 16665880 ns.
It does not include more than 30 fps in aeAvailableTargetFpsRanges, so I can imagine that auto exposure would have to be turned off.

**Comment by pedroSG94** [comment]
Hello,

First of all, keep in mind that no matter if your device support record with 60fps using default camera app or not because not all camera features are exposed to camera2 API

As you said the fps supported by your device using camera2 is managed here:
https://github.com/pedroSG94/RootEncoder/blob/master/encoder/src/main/java/com/pedro/encoder/input/video/Camera2ApiManager.java#L263

I think that it is the only way to get the fps supported by your device using camera2. If you can find other way to get the fps supported by the device and set it correctly, let me know it and I will implement it.
Also, you can try skip that method and set your fps range directly here (you can replace that line):
https://github.com/pedroSG94/RootEncoder/blob/master/encoder/src/main/java/com/pedro/encoder/input/video/Camera2ApiManager.java#L259

**Reply by Ernovation** [reply]
Hello Pedro,

I agree with your remarks.

My question was not specifically about RootEncoder, but more about experiences of other users with 60 fps. I was curious if other users were able to stream or record with 60 fps, because their phone supports it.

It is not easy to find information about supported frame rates for Camera2 for different phones, so I thought I perhaps could start a "discussion" about it here.

And perhaps a higher frame rate would be possible if you turn off auto exposure, because aeAvailableTargetFpsRanges might give frame rates for which auto exposure is possible. That could be an interesting experiment, but it would not help for my S23 Ultra, because that has minFrameDuration that does not support 30 fps.

**Comment by pedroSG94** [comment]
Hello,

The fps also depend of the resolution. For example with my Pixel 6a using 1280x720 I have this fps supported:
```
15 - 15
15 - 30
30 - 30
15 - 60
60 - 60
```
But using 640x480 I have this result:
```
15 - 15
15 - 30
30 - 30
```
I used this code to check it:
```java
    for (Range r: rtmpCamera2.getSupportedFps(new Size(1280, 720), CameraHelper.Facing.BACK)) {
      Log.e("Pedro", r.getLower() + " - " + r.getUpper());
    }
```

Using a resolution that support 60fps and set 60 fps in preparevideo is working for me using pixel 6a. It is working with 720p, 1080p and 4k resolutions with my Pixel 6a.

About the reason because 640x480 doesn't support 60fps. I think that this resolution is not supported by the camera hardware directly so the camera is configured to open other resolution and crop image so the camera is capped to this frame rate for this case.

**Reply by Ernovation** [reply]
Thanks, I will consider a Pixel phone for my next phone. 
Too bad that it is not easy to find these specifications for different phones. I believe a public database with these specifications for different phones would be helpful for people like me, who would like to use their phones for e.g. live video streaming.
Just for fun I wrote some quick and dirty code to log the entire CameraCharacteristics for the different cameras in a device : 


```
var cameraManager = (CameraManager) this.getApplicationContext().getSystemService(Context.CAMERA_SERVICE);

try {
    var cameras = cameraManager.getCameraIdList();
    Log.i(this.getClass().getSimpleName(), "**\r\n**\r\nCamera info for " + Build.MANUFACTURER + " " + Build.MODEL);
    for (var cameraId : cameras) {
        var info = "\r\n**\r\nInfo for camera with id=" + cameraId + ":\r\n**\r\n";
        var characteristics = cameraManager.getCameraCharacteristics(cameraId);
        for (var c : characteristics.getKeys()) {
            info = info + c.getName() + "=" + format(characteristics.get(c)) + "\r\n";
        }
        Log.i(this.getLocalClassName(), info);
    }
} catch (CameraAccessException e) {
    throw new RuntimeException(e);
}
```
and 
```
    private String format(long[] l) {
        if (l.length == 0) {
            return "(long[])[]";
        }
        var s = "(long[])[" + l[0];
        for (int i = 1; i < l.length; i++) {
            s += "," + l[i];
        }
        s += "]";
        return s;
    }

    private String format(int[] l) {
        if (l.length == 0) {
            return "(int[])[]";
        }
        var s = "(int[])[" + l[0];
        for (int i = 1; i < l.length; i++) {
            s += "," + l[i];
        }
        s += "]";
        return s;
    }

    private String format(float[] l) {
        if (l.length == 0) {
            return "(float[])[]";
        }
        var s = "(float)[" + l[0];
        for (int i = 1; i < l.length; i++) {
            s += "," + l[i];
        }
        s += "]";
        return s;
    }

    private String format(boolean[] l) {
        if (l.length == 0) {
            return "(boolean[])[]";
        }
        var s = "(boolean[])[" + l[0];
        for (int i = 1; i < l.length; i++) {
            s += "," + l[i];
        }
        s += "]";
        return s;
    }

    private String format(Object[] l) {
        if (l.length == 0) {
            return "(" + l.getClass().getSimpleName() + ")[]";
        }
        var s = "(" + l.getClass().getSimpleName() + ")[" + format(l[0]);
        for (int i = 1; i < l.length; i++) {
            s += "," + format(l[i]);
        }
        s += "]";
        return s;
    }

    private String format(StreamConfigurationMap m) {
        var formats = m.getOutputFormats();
        var res = "(StreamConfigurationMap)\r\n";
        for (var f: formats) {
            res = res + "- format " + f + ":\r\n";
            var sizes = m.getOutputSizes(f);
            for (var s: sizes) {
                var duration = m.getOutputMinFrameDuration(f, s);
                var stall = m.getOutputStallDuration(f, s);
                res = res + "  - size " + s + ": \tduration=" + duration + "; \tstall=" + stall + "\r\n";
            }
        }
        return res;

    }

    private String format(MandatoryStreamCombination c) {
        var res = "(MandatoryStreamCombination)\r\n";
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            res += "- description=\"" + c.getDescription() + "\"";
            for (var info : c.getStreamsInformation()) {
                res += "\r\n  StreamsInformation:";
                res += "\r\n  - format=" + info.getFormat();
                res += "\r\n  - isInput=" + info.isInput();
                res += "\r\n  - sizes=";
                for (var s : info.getAvailableSizes()) {
                    res += s + ";";
                }
                if (Build.VERSION.SDK_INT >= 33) {
                    res += "\r\n  - streamUseCase=" + info.getStreamUseCase();
                    res += "\r\n  - 10BitCapable=" + info.is10BitCapable();
                    if (info.is10BitCapable()) {
                        res += "\r\n  - 10BitFormat=" + info.get10BitFormat();
                    }
                    res += "\r\n  - isMaximumSize=" + info.isMaximumSize();
                    res += "\r\n  - isUltraHighResolution=" + info.isUltraHighResolution();
                }
                res += "\r\n";
            }
        }

        return res;
    }
    
    private String format(Object o) {
        if (o instanceof int[]) {
            return format((int[])o);
        } else if (o instanceof long[]) {
            return format((long[])o);
        } else if (o instanceof boolean[]) {
            return format((boolean[])o);
        } else if (o instanceof float[]) {
            return format((float[])o);
        } else if (o instanceof Object[]) {
            return format((Object[])o);
        } else if (o instanceof StreamConfigurationMap) {
            return format((StreamConfigurationMap)o);
        } else if (Build.VERSION.SDK_INT >= 29 && o instanceof MandatoryStreamCombination) {
            return format((MandatoryStreamCombination) o);
        }

        return "(" + o.getClass().getSimpleName() + ")" + o.toString();
    }
```
I have attached the resulting log files for my Galaxy S23 Ultra and my Galaxy S10, for those who are interested.
[S23Ultra.log](https://github.com/pedroSG94/RootEncoder/files/14089453/S23Ultra.log)
[S10.log](https://github.com/pedroSG94/RootEncoder/files/14089455/S10.log)

**Comment by ramaatamai** [comment]
Hi @pedroSG94 I am using Samsung S10 device supports 60fps in UHD,FHD frame rates in the built-in camera app.When i i want to implement camera application using camera2 api it doesn't support 60fps .Can you tell me the reason?

**Reply by pedroSG94** [reply]
Hello,

This is common. 
Most manufacturers hide camera features to android camera APIs and that features are only available on the native camera app.

I think that this is your case.

**Comment by ramaatamai** [comment]
@pedroSG94 Thank for the reply
